Add FXIOS-11526 [Homepage] [Telemetry] event for tapping on items

cyndichin opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-11526)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/25086)

## :bulb: Description
Add telemetry event for tapping on an item in the homepage section. This includes the following items:

**Card Items**
- top sites
- jump back in local tab
- jump back in synced tab
- bookmark item
- pocket item
- pocket discover item

**Other**
- bookmarks show all
- jump back in show all (local tabs)
- jump back in synced tabs show all
- customize homepage

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

cyndichin commented:
Waiting on telemetry naming guidelines

cyndichin commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -2734,6 +2734,26 @@ homepage:
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/11526
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/18518
```
```suggestion
      - https://github.com/mozilla-mobile/firefox-ios/pull/25297
```

cyndichin commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -2734,6 +2734,26 @@ homepage:
+      item_name:
+       type: string
+       description: |
+        The name associated with the item that was tapped on the homepage.
```
@ih-codes Revisiting this PR, instead of section name, I am using item name. In `telemetryLabel`, I have separated out synced tab and pocket discover more item. Any issues with leaving it `item_name` or do you prefer `item_type`?

ih-codes replied:
Hmmmm I think `item_type` maybe? Since it's more like a category rather than an individual name in my mind. 🤔

cyndichin commented:
@Mergifyio backport release/v138

cyndichin commented:
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

We want to know whether users are interested in certain items on the homepage by tapping and navigating to the respective links.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:

We will learn whether certain sections are valuable to users and we can let us know where to spend our efforts in improving the homepage.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

No alternative methods as these new events give us the sufficient data to the questions above while following the correct telemetry naming convention for consistency. 

4) Can current instrumentation answer these questions?

Yes for the old homepage, we have existing telemetry events, but will be replacing those old telemetry events with these new ones.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>homepage.card_item_tapped</td>
    <td>2</td>
    <td>FXIOS-11473</td>
  </tr>
  <tr>
    <td>settings.main_menu.option_selected</td>
    <td>2</td>
    <td>FXIOS-11473</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/

7) How long will this data be collected?  Choose one of the following:

TBD

8) What populations will you measure?

TBD

9) If this data collection is default on, what is the opt-out mechanism for users?

User has the general option of disable telemetry or/and studies.

10) Please provide a general description of how you will analyze this data.

We will analyze aggregated information: items tapped from the homepage and the most common items that were tapped.

11) Where do you intend to share the results of your analysis?

Mozilla Internal (Product and Dev Team).

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:

No. We will collect the info using Glean. We will analyze the data with the traditional Looker/SQL database that Mozilla uses for all other telemetry.

cyndichin commented:
Still need to confirm data collection review details with John.

ih-codes commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -2734,6 +2734,26 @@ homepage:
+      item_name:
+       type: string
+       description: |
+        The type of item that belongs to on the homepage. This name is found in `HomepageDiffableDataSource.HomeItem.telemetryLabel`.
```
```
The type of item that belongs to on the homepage.
```
I think maybe a word missing here?

ih-codes commented:
@cyndichin After discussing with Andy, here is what we came up with:
```
homepage.item_tapped   
   extra: section (e.g. top sites, bookmarks, jump back in, stories…)  
   extra: type (item, or show_all_button for "show all" buttons, etc.)
```
So an item in each section would just be "item" for what we were calling the "cards" the other day. The `type` becomes more specific if we tap something like one of the "Show all" buttons, the customize button, the privacy toggle, etc.

If we were to tap the privacy toggle, we would do something like this:
```
homepage.item_tapped   
   extra: section (e.g. header)  
   extra: type (e.g. private_mode_toggle)
```

Or the customize homepage button at the bottom:
```
homepage.item_tapped   
   extra: section (e.g. customize_homepage)  -- or use "footer" if that's what we have in code
   extra: type (e.g. customize_homepage_button)
```

cyndichin commented:
> @cyndichin After discussing with Andy, here is what we came up with:
> 
> ```
> homepage.item_tapped   
>    extra: section (e.g. top sites, bookmarks, jump back in, stories…)  
>    extra: type (item, or show_all_button for "show all" buttons, etc.)
> ```
> 
> So an item in each section would just be "item" for what we were calling the "cards" the other day. The `type` becomes more specific if we tap something like one of the "Show all" buttons, the customize button, the privacy toggle, etc.
> 
> If we were to tap the privacy toggle, we would do something like this:
> 
> ```
> homepage.item_tapped   
>    extra: section (e.g. header)  
>    extra: type (e.g. private_mode_toggle)
> ```
> 
> Or the customize homepage button at the bottom:
> 
> ```
> homepage.item_tapped   
>    extra: section (e.g. customize_homepage)  -- or use "footer" if that's what we have in code
>    extra: type (e.g. customize_homepage_button)
> ```

Thanks! I made some of the updates, I changed `type` to `item_type` to avoid confusion with the data type. I used the underscore instead of spaces (but both were referenced so not too sure what's preferred).

In addition to the card items, I also included tapping on the show alls and customize button. The privacy toggle will be removed in a separate PR.

Note: wallpaper and message section has there own telemetry that I will leave as is. I renamed pocket to stories / story. The previous pocket telemetry had a position int associated. Do we want to include that?

cyndichin commented on firefox-ios/firefox-ios-tests/Tests/XCUITests/AddressesTests.swift:
```diff
@@ -3,6 +3,7 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
 import Foundation
+import XCTest
```
```suggestion```

cyndichin commented:
Status Update - PR is ready for review, need to confirm data collection form with John (meeting on Tuesday)

ih-codes reviewed (approved):
A tiny naming nit, otherwise this looks good to me! Great unit tests too! 👏

cyndichin commented on firefox-ios/Client/Telemetry/HomepageTelemetry.swift:
```diff
@@ -6,12 +6,45 @@ import Foundation
 struct HomepageTelemetry {
-    private let gleanWrapper: GleanWrapper
+    enum TappedItemType: String {
+        case topSite = "top_site"
```
Thanks for testing! I plan to add these to the spreadsheet + have QA go through it to test.

cyndichin commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -2715,6 +2715,32 @@ firefox_home_page:
+       description: |
+        The type of item that was tapped on the homepage. This name is found in `HomepageTelemetry.TappedItemType`.
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/11526
```
```suggestion
      - https://github.com/mozilla-mobile/firefox-ios/issues/25297
```

cyndichin commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -2715,6 +2715,32 @@ firefox_home_page:
+       description: |
+        The type of item that was tapped on the homepage. This name is found in `HomepageTelemetry.TappedItemType`.
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/issues/25297
```
```suggestion
      - https://github.com/mozilla-mobile/firefox-ios/issues/25086
```
